Refactor gitlab modules

Lunik opened the pull request:
##### SUMMARY
Refactoring Gitlab module to fix Inconsistency between gitlab modules #51022

##### ISSUE TYPE
- Bugfix Pull Request
- Feature Pull Request

##### COMPONENT NAME
gitlab_user
gitlab_group
gitlab_deploy_key
gitlab_hooks
gitlab_project

##### ADDITIONAL INFORMATION
This PR fixes #29310.
This PR fixes #32836.
This PR fixes #34255.
This PR fixes #40767.
This PR fixes #41347.
This PR fixes #44564.
This PR fixes #44677.
This PR fixes #50309.
This PR fixes #51022.

Commit:
gitlab_group: refactor module

Commit:
gitlab_user: refactor module

Commit:
gitlab_group, gitlab_user; pylint

Commit:
gitlab_project: refactor module

Commit:
gitlab_group, gitlab_project, gitlab_user: Enchance modules

- Add generic loop to update object
- Enchance return messages
- PyLint

Commit:
gitlab_runner: refactor module

Commit:
gitlab_hooks: refactor module

Commit:
gitlab_deploy_key: refactor module

Commit:
gitlab_group: enchance module and documentation

- Enchange function arguments
- Add check_mode break
- Rewrite module documentation

Commit:
gitlab_hook: enchance module and documentation

- Rewrite documentation
- Enchance function parameters
- Rename functions

waheedi commented on lib/ansible/modules/source_control/gitlab_deploy_key.py:
```diff
@@ -1,40 +1,41 @@
-# (c) 2018, Marcus Watkins <marwatk@marcuswatkins.net>
-# Based on code:
-# (c) 2013, Phillip Gentry <phillip@cx.com>
+# (c) 2015, Werner Dijkerman (ikben@werner-dijkerman.nl)
```
I'm not sure about copy rights updates, but I think it would be good to describe why these updates are needed?

Lunik replied:
wrong copy/past

waheedi commented:
Great progress @Lunik , It would be good to squash these commit with a clear commit messages, also it would have been better to do these refactors after the pull requests have been migrated/reviewed/rejected so no overlaps would occur.

Also I have commented on one of the commits, regarding copy rights changes.

Commit:
gitlab_project: enchance module and documentation

- Rewrite documentation
- Enchance function parameters
- Add try/except on project creation

Commit:
gitlab_runner: enchance module and documentation

- Rewrite documentation
- Fix Copyright
- Enchance function arguments
- Add check_mode break
- Add missing function: deletion

Commit:
gitlab_user: enchance module and documentation

- Rewrite documentation
- Enchance function parameters
- Add check_mode break
- Add try/except on user creation

Commit:
gitlab_deploy_key, gitlab_group, gitlab_hooks, gitlab_project,
gitlab_runner, gitlab_user: Fix residual bugs

- Fix Copyright
- Fix result messages
- Add missing check_mode break

Lunik commented:
> Great progress @Lunik , It would be good to squash these commit with a clear commit messages, also it would have been better to do these refactors after the pull requests have been migrated/reviewed/rejected so no overlaps would occur.
> 
> Also I have commented on one of the commits, regarding copy rights changes.

Are you talking about #50442 pull request ? Because, I should close it

Lunik commented:
Just Squashed and renamed commit 👍

marwatk commented:
Looking good. Don't forget to add aliases to the old keys and I think you can remove the [gitlab helper from module_utils](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/gitlab.py).

Lunik commented:
Unit tests of `gitlab_deploy_key`, `gitlab_hooks` and `gitlab_project` failed because of the full refactoring of the module.
What should I do ?

Commit:
gitlab_deploy_key, gitlab_group, gitlab_hooks, gitlab_project, gitlab_runner, gitlab_user: pylint

Lunik commented:
> Unit tests of `gitlab_deploy_key`, `gitlab_hooks` and `gitlab_project` failed because of the full refactoring of the module.
> What should I do ?

Actually, they can't be reused because of the modification of the way that the module communicate with the Gitlab instance. It doesn't make direct call to the API, now it use a python library that do the job. So using a pytest mocker to test the module won't work.

Commit:
gitlab_runner: Add substitution function for 'cmp' in python3

Commit:
unit-test: remove deprecated gitlab module tests

- gitlab_deploy_key
- gitlab_hooks
- gitlab_project

Actually, they can't be reused because of the modification of the way that the module communicate with the Gitlab instance. It doesn't make direct call to the API, now it use a python library that do the job. So using a pytest mocker to test the module won't work.

marwatk commented on lib/ansible/modules/source_control/gitlab_deploy_key.py:
```diff
@@ -9,35 +7,46 @@
+    description:
+      - Gitlab password for login_user
+    version_added: "2.8"
+  login_token:
```
Can we add an alias for `private_token` for the modules that used to accept it? (Also applies to `gitlab_hooks`)

Lunik replied:
will be fixed in next commits

marwatk commented on lib/ansible/modules/source_control/gitlab_deploy_key.py:
```diff
@@ -9,35 +7,46 @@
   project:
     description:
-      - Numeric project id or name of project in the form of group/name
+      - Full path of project in the form of group/name
```
It seems like python-gitlab [supports numeric ids](https://python-gitlab.readthedocs.io/en/stable/gl_objects/projects.html), any reason to limit the docs here?

Lunik replied:
I will create a function in gitlab library to allow id again

marwatk commented on lib/ansible/modules/source_control/gitlab_deploy_key.py:
```diff
@@ -9,35 +7,46 @@
 version_added: "2.6"
 options:
-  api_url:
+  server_url:
```
This is going to break old users, think it's worth also accepting `api_url`? Or potentially using an alias and stripping off `/api`? That feels hackish, though. (also applies to `gitlab_hooks`)

Lunik replied:
will be fixed in next commits

marwatk commented on lib/ansible/modules/source_control/gitlab_deploy_key.py:
```diff
@@ -94,140 +102,225 @@
+    @param full_path Full path of the project
+    '''
+    def findProject(self, name, full_path):
+        projects = self._gitlab.projects.list(search=name, as_list=False)
```
I think this answers my question on the numeric id thing, but why not just call `self._gitlab.projects.get(full_path)`?

Lunik replied:
I will create a function in gitlab library

marwatk commented on lib/ansible/modules/source_control/gitlab_project.py:
```diff
@@ -151,141 +165,132 @@
+    @param name Name of the groupe
+    @param full_path Complete path of the Group including parent group path. <parent_path>/<group_path>
+    '''
+    def findGroup(self, name, full_path):
```
`findGroup` and `findProject` are repeated in a few places, worth moving out into `module_utils`? There's a `gitlab` library out there already that won't be needed anymore, you could use that for generic stuff.

Lunik replied:
I will create a function in gitlab library

marwatk commented on lib/ansible/modules/source_control/gitlab_runner.py:
```diff
@@ -31,413 +28,348 @@
-        required: False
-        default: ["docker"]
-        type: list
+  login_token:
```
Alias here for `private_token`?

Lunik replied:
will be fixed in next commits

marwatk reviewed (changes requested):
Looks great, don't forget to remove `module_utils/gitlab.py` if you don't use it for any reusable code.

Commit:
gitlab_deploy_key, gitlab_group, gitlab_hooks, gitlab_project, gitlab_runner, gitlab_user: add copyright

Commit:
gitlab_deploy_key, gitlab_group, gitlab_hooks, gitlab_project, gitlab_runner, gitlab_user: Support old parameters format

Lunik commented:
ready_for_review

dagwieers commented:
@Lunik Can you remove the references in those files as indicated ? We can only accepts PRs if they succeed on all tests. (Which may mean updating other files like BOTMETA.yml or ignore files)

dagwieers commented on lib/ansible/module_utils/gitlab.py:
```diff
@@ -3,6 +3,7 @@
 # (c) 2018, Marcus Watkins <marwatk@marcuswatkins.net>
```
Format of copyright statements is now different.
```suggestion
# Copyright: (c) 2018, Marcus Watkins <marwatk@marcuswatkins.net>
```

Lunik replied:
Will be fix in the next commit

dagwieers commented on lib/ansible/modules/source_control/gitlab_deploy_key.py:
```diff
@@ -1,43 +1,57 @@
+  server_url:
     description:
-      - GitLab API url, e.g. https://gitlab.example.com/api
+      - Url of Gitlab server, with protocol (http or https).
```
Use URL instead of Url.
```suggestion
      - The URL of the Gitlab server, with protocol (i.e. http or https).
```

Lunik replied:
Will be fix in the next commit

dagwieers commented on lib/ansible/modules/source_control/gitlab_deploy_key.py:
```diff
@@ -1,43 +1,57 @@
+    description:
+      - When using https if SSL certificate needs to be verified.
+    type: bool
+    default: 'yes'
```
These days we don't expect this to be strings, we now generate documentation correctly. So this can be real booleans.
```suggestion
    default: yes
```

Lunik replied:
Will be fix in the next commit

dagwieers commented on lib/ansible/modules/source_control/gitlab_deploy_key.py:
```diff
@@ -1,43 +1,57 @@
+    version_added: "2.8"
+    aliases:
+      - api_url
+  verify_ssl:
```
This needs to be C(validate_certs) which is the standardized parameter name.

Lunik replied:
Will be fix in the next commit

dagwieers commented on lib/ansible/modules/source_control/gitlab_deploy_key.py:
```diff
@@ -1,43 +1,57 @@
+    default: 'yes'
+    aliases:
+      - validate_certs
+      - enable_ssl_verification
```
Please don't add any aliases except for backward compatibility. We don't want to offer more choice than necessary.

Lunik replied:
Will be fix in the next commit

dagwieers commented on lib/ansible/modules/source_control/gitlab_deploy_key.py:
```diff
@@ -94,140 +107,216 @@
+        self.deployKeyObject = deployKey
+        if changed:
+            if self._module.check_mode:
+                self._module.exit_json(changed=True, result="DeployKey should have been updated.")
```
Please use 'msg' for returned messages, this is a standardized return value.

Lunik replied:
Will be fix in the next commit

dagwieers commented on lib/ansible/modules/source_control/gitlab_deploy_key.py:
```diff
@@ -94,140 +107,216 @@
-            access_token=dict(required=False, no_log=True),
-            private_token=dict(required=False, no_log=True),
+            server_url=dict(required=True, aliases=['api_url']),
+            verify_ssl=dict(required=False, default=True, type='bool', aliases=['enable_ssl_verification', 'validate_certs']),
```
Same here, default parameter is **validate_certs**, the aliases should be removed.

We also simplify this by *not* adding `required=False`, and always starting with the type.
Since every parameter has a type, starting with this adds consistency.

Lunik replied:
Will be fix in the next commit

dagwieers commented on lib/ansible/modules/source_control/gitlab_deploy_key.py:
```diff
@@ -94,140 +107,216 @@
-            private_token=dict(required=False, no_log=True),
+            server_url=dict(required=True, aliases=['api_url']),
+            verify_ssl=dict(required=False, default=True, type='bool', aliases=['enable_ssl_verification', 'validate_certs']),
+            login_user=dict(required=False, no_log=True),
```
So removed `required=False` and add `type='str' instead. Do this for the others as well.

Lunik replied:
Will be fix in the next commit

dagwieers commented on lib/ansible/modules/source_control/gitlab_deploy_key.py:
```diff
@@ -94,140 +107,216 @@
+            login_user=dict(required=False, no_log=True),
+            login_password=dict(required=False, no_log=True),
+            login_token=dict(required=False, no_log=True, aliases=['private_token']),
+            state=dict(default="present", choices=["present", 'absent']),
```
So add `type='str'` here too. And we tend to sort lists if the order is of no importance.

Lunik replied:
Will be fix in the next commit

dagwieers commented on lib/ansible/modules/source_control/gitlab_deploy_key.py:
```diff
@@ -94,140 +107,216 @@
+            ['login_user', 'login_token']
         ],
-        supports_check_mode=True,
+        supports_check_mode=True
```
Leave the trailing comma. This is explicitly allowed by python in order to make line-modifications (like moving the order of lines, or adding lines) without having to update unrelated lines.

Lunik replied:
Will be fix in the next commit

dagwieers commented on lib/ansible/modules/source_control/gitlab_deploy_key.py:
```diff
@@ -94,140 +107,216 @@
-        module.exit_json(changed=changed, msg='Success', result=response, previous_version=existing)
-    else:
-        module.fail_json(msg='Failure', error=response)
+            module.exit_json(changed=True, result="Successfully created or updated the deployKey %s" % key_title,
```
Please use `msg` instead of `result`. Also the standard casing for both parameters as well as return values is snail_case. So it would become `deploy_key` instead of `deployKey`.

Lunik replied:
Will be fix in the next commit

dagwieers commented on lib/ansible/modules/source_control/gitlab_group.py:
```diff
@@ -15,59 +15,69 @@
-            - If not supplied, the group_name will be used.
+  server_url:
+    description:
+      - Url of Gitlab server, with protocol (http or https).
```
Same remarks here, use URL instead of Url.
```suggestion
      - The URL of the Gitlab server, with protocol (i.e. http or https).
```

Lunik replied:
Will be fix in the next commit

dagwieers reviewed (changes requested):
Various changes are required for all the modules.
Also:
- You remove the unit tests, but you don't add any new unit tests, so we are now testing less.
- Especially since we don't have any integration tests, it would be nice to include those in order to verify (individually) that the functionality is working as designed on everyone's infrastructure.

Lunik committed:
module_utils Gitlab: Edit copyright

Lunik committed:
gitlab_deploy_key, gitlab_group, gitlab_hooks, gitlab_project,
gitlab_runner, gitlab_user: Unifying module inputs

- Rename verify_ssl into validate_certs to match standards
- Remove unused alias parameters
- Unify parameters type and requirement
- Reorder list order

Lunik committed:
gitlab_deploy_key, gitlab_group, gitlab_hooks, gitlab_project, gitlab_runner, gitlab_user: Unifying module outputs

- Use standard output parameter "msg" instead of "return"
- Use snail_case for return values instead of camelCase

Lunik committed:
validate-module: remove sanity ignore

Lunik commented:
> The test `ansible-test sanity --test botmeta` [[explain](https://docs.ansible.com/ansible/devel/dev_guide/testing/sanity/botmeta.html)] failed with 1 error:
> 
> ```
> .github/BOTMETA.yml:0:0: Can't find 'test/units/modules/source_control/test_gitlab_.*' in this branch
> ```
> [click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

Don't know how to fix this one ... 🤔

dagwieers commented:
Remove the reference to that file from the file *.github/BOTMETA.yml* and commit that change in your PR.

Lunik committed:
BOTMETA: remove gitlab_* test

- This tests need to be completely rewriten because of the refactoring
of these modules
- TodoList Community Wiki was updated

Lunik commented:
> Remove the reference to that file from the file _.github/BOTMETA.yml_ and commit that change in your PR.

Just rebase the branch, the line popped in the file 🎉

dagwieers commented:
Oh right, you should always rebase your branch ;-)

Shaps commented:
@Lunik I'm adding integration tests based on your changes, will report back if I find any issues so once we add them we're confident modules will pass tests.

Lunik commented:
@Shaps I'm not really comfortable with testing, thank's for taking this job 👍

Lunik commented:
> @Lunik You really need to learn how to add integration tests, as this is the best way to find issues with changes you make. Things may seem to work when you test it while developing, but it's not guaranteed to work correctly if it isn't testing all code paths.

Yep I agree, I'm working on it to improve my skills 😄 But for now, pytest & cie are just black magic for me...

Shaps commented:
@Lunik Integration tests are just standard ansible roles with some `asserts` here and there ;)

Lunik commented:
Is there a way to make a `module_utils` for unit tests ? Like we have for modules ?
The main purpose is to store mock function.

Lunik committed:
gitlab_user: Fix group identifier

Lunik commented:
I have tried to write my first unit test for the module `gitlab_user` [source_code](https://gist.github.com/Lunik/f1b70feb89f9b40d9ff8fa26d484a5b2).
Am I doing it right ?

dagwieers commented:
Questions about unit tests are best directed to @mattclay

Shaps commented:
Do we really need unit tests for the modules? I think we can cover most of this using the ansible-test integration tests?

Lunik committed:
gitlab modules: Add units tests

Lunik commented:
> Do we really need unit tests for the modules? I think we can cover most of this using the ansible-test integration tests?

Agree, there is not a lot of "intelligent" code in those modules. It's just a lot of API call's.
With unit tests, I can mock this calls so I'm re-testing the `python-gitlab` module.

In my opinion, if we have special function with data processing, then we should make unit tests for those functions. In the other case, we should use integration tests

Lunik commented:
About `gitlab_hooks`:
As mentioned in https://github.com/ansible/ansible/pull/39991#issuecomment-459241567, we should maybe migrate `hooks` events options (push_events, issues_events, ...) into a list parameter (`events` or something like that) and create a deprecation warning on those parameters.

dagwieers commented:
Unit tests are nice to have and are complementary to integration tests. So ideally we have both.

@Lunik I noticed you are force-pushing a lot, which makes it harder for people to re-review your changes (as it is unclear what you changed). And before we can review and merge we need to be sure that you consider your PR final and complete (if not, add WIP in the PR subject), otherwise we are running in circles. So let's review, test and merge this PR ASAP.

Lunik commented:
> Unit tests are nice to have and are complementary to integration tests. So ideally we have both.
> 
> @Lunik I noticed you are force-pushing a lot, which makes it harder for people to re-review your changes (as it is unclear what you changed). And before we can review and merge we need to be sure that you consider your PR final and complete (if not, add WIP in the PR subject), otherwise we are running in circles. So let's review, test and merge this PR ASAP.

Ok let's add back WIP, push back integration tests that I have made and I will notify the team wen done. No more push-force I will squash commit when everything is ready to ship.

Shaps commented:
@Lunik Integration tests ( even though basic-ish ) are already part of #51490. Once it gets merged, you can use those to test the modules. I'll write up a wiki page on how I setup my environment to run the tests.

Lunik committed:
gitlab_project: Fix when group was empty

Lunik committed:
gitlab_deploy_key: edit return msg

Lunik committed:
module_utils gitlab: fall back to user namespace is project not found

Lunik commented:
Just run all integration tests from #51490 and some other more. Everything is green.
I will add unit tests as soon as I can.

Lunik committed:
unit test: gitlab module fake current user

Lunik committed:
gitlab_user: fix access_level verification

Lunik committed:
gitlab unit tests: use decoration instead of with statement

Lunik commented:
Unit tests are failing with python 2.6 because  `python-gitlab` library is not compatible with python 2.6 anymore.

Community thread : https://github.com/ansible/community/issues/431#issuecomment-459488720

dagwieers commented:
You will have to test for Python 2.6 and avoid running the tests there. cc @mattclay

Lunik commented:
> You will have to test for Python 2.6 and avoid running the tests there.

@dagwieers Sorry, I didn't get what you mean. Can you be more specific ?

dagwieers commented:
@Lunik If you want your unit-tests to succeed on the Python 2.6 environment, you have to catch when you are running Python 2.6 and skip the tests (so they don't fail).
```python
        # Python 2.7+ is needed for python-gitlab
        if sys.version_info < (2, 7):
            return
```

Lunik committed:
unit tests: gitlab module skip python 2.6

Lunik committed:
unit tests: gitlab module skip library import if python 2.6

Lunik committed:
gitlab unit tests: use builtin unittest class

Lunik committed:
gitlab unit tests: use custom test class

Lunik committed:
unit test: gitlab module lint

mattclay commented on test/units/utils/test_gitlab.py:
```diff
@@ -0,0 +1,559 @@
+# Copyright: (c) 2019, Guillaume Martinez (lunik@tiwabbit.fr)
```
Move this to `test/units/modules/source_control/gitlab.py`. This will make sure that the correct unit tests are executed when changes are made to the file.

Lunik replied:
resolved with dcdd770 and 9154899

Lunik committed:
unit tests: move gitlab utils

Lunik committed:
unit test: gitlab fix imports

Lunik committed:
gitlab_module: edit requirement

python-gitlab library require python >= 2.7

Lunik commented:
Should I add myself as an Author on those modules ? 🤔

Lunik committed:
gitlab_module: add myself as author

Lunik commented:
Shippable job failed for no reason... [105465#85](https://app.shippable.com/github/ansible/ansible/runs/105465/85/console)

Shaps commented:
@Lunik CI is all green now. The easiest way to get those unrelated failures fixed is to get on `#ansible-devel` in freenode IRC and ask for the specific failed tests to be restarted.

dagwieers commented on lib/ansible/modules/source_control/gitlab_deploy_key.py:
```diff
@@ -1,43 +1,60 @@
      - Adds, updates and removes project deploy keys
 version_added: "2.6"
+author:
+  - "Marcus Watkins (@marwatk)"
```
There is no need to add quotes here.
Please remove them from all modules.

Lunik replied:
Fixed by 68c42b3

dagwieers commented on lib/ansible/modules/source_control/gitlab_deploy_key.py:
```diff
@@ -1,43 +1,60 @@
+  - python-gitlab python module
 options:
-  api_url:
+  server_url:
```
So, looking at our existing list of modules, **api_url** is more popular than **server_url**.
I guess we need to make sure we are making the right changes wrt. parameter naming.
```
[dag@moria ansible.git]$ grep -rl 'api_url' lib/ansible/modules | grep '\.py$' | wc -l
77
[dag@moria ansible.git]$ grep -rl 'server_url' lib/ansible/modules | grep '\.py$' | wc -l
21
```

dagwieers replied:
In fact, it seems to be standardized already in *lib/ansible/module_utils/api.py* as **api_url**.
cc @bcoca

Lunik replied:
Well in that case, I will change `server_url` to `api_url` on all modules

dagwieers commented on lib/ansible/modules/source_control/gitlab_deploy_key.py:
```diff
@@ -1,43 +1,60 @@
+  server_url:
     description:
-      - GitLab API url, e.g. https://gitlab.example.com/api
+      - The URL of the Gitlab server, with protocol (i.e. http or https).
```
Please add parameter types to every parameter. This ensures the module documentation includes parameter types.
```suggestion
      - The URL of the Gitlab server, with protocol (i.e. http or https).
    type: str
```

Lunik replied:
Fixed by 68c42b3

dagwieers commented on lib/ansible/modules/source_control/gitlab_deploy_key.py:
```diff
@@ -51,41 +68,39 @@
       - When C(absent) it will be removed from the project if it exists
     required: true
-    default: present
+    default: "present"
```
No need for quotes here. We tend to only quote where it is required by YAML.
This ensures that our users understand when quoting is actually required.

Lunik replied:
Fixed by 68c42b3

dagwieers commented on lib/ansible/modules/source_control/gitlab_deploy_key.py:
```diff
@@ -51,41 +68,39 @@
     required: true
-    default: present
+    default: "present"
     choices: [ "present", "absent" ]
```
No quotes are required here either.
We tend to sort lists alphabetically in those cases where the order is of no importance.
That's in general a good practice

Lunik replied:
Fixed by 68c42b3

dagwieers commented on lib/ansible/modules/source_control/gitlab_deploy_key.py:
```diff
@@ -51,41 +68,39 @@
 '''
 
 EXAMPLES = '''
 # Example adding a project deploy key
```
We expect all examples to have a `name:` entry (which in this case could be the comment just above the example).

Lunik replied:
Fixed by 68c42b3

dagwieers commented on lib/ansible/modules/source_control/gitlab_group.py:
```diff
@@ -172,46 +281,65 @@ def main():
-
     server_url = module.params['server_url']
-    validate_certs = module.params['validate_certs']
+    verify_ssl = module.params['validate_certs']
```
I would rather keep this as validate_certs, as this is the actional parameter name as well.
Besides using verify_ssl and having the option ssl_verify adds to the confusion.

Lunik replied:
fixed by e09d583

dagwieers commented on lib/ansible/modules/source_control/gitlab_hooks.py:
```diff
@@ -1,43 +1,61 @@
 #!/usr/bin/python
-# -*- coding: utf-8 -*-
```
This should stay in each file to ensure the file is considered UTF-8 by Python.
Don't remove those lines.

Lunik replied:
Fixed with 701c124

dagwieers reviewed (changes requested):
Various changes.

dagwieers commented:
The Gitlab modules can share documentation fragments (see *lib/ansible/plugins/doc_fragments/*) and can share argspec fragments (see *lib/ansible/module_utils/api.py*). Plenty of examples :-)

Lunik committed:
gitlab_modules: add python encoding tag

Lunik committed:
gitlab_modules: keep consistency between variable name "validate_certs"

Lunik committed:
gitlab_modules: enchance documentation

Lunik commented:
> The Gitlab modules can share documentation fragments (see _lib/ansible/plugins/doc_fragments/_) and can share argspec fragments (see _lib/ansible/module_utils/api.py_). Plenty of examples :-)

This fragments documentation seems great ! But options description can't be modified to be more specific like "The URL of the Gitlab server, with protocol (i.e. http or https).".

Lunik committed:
gitlab_runner: fix syntax error in documentation

dagwieers commented:
> No backward compatibilities for parameters.

You are right, it might be a reason to not share this. On the other hand you could leave the aliases in all modules so that they are more alike (including the aliases).

PS We still need to add some library functions for deprecating aliases over time.

Lunik commented:
Yep, I will try something tomorrow. We will see

Lunik committed:
gitlab_module: use basic_auth module_utils and add deprecation warning

Lunik committed:
gitlab_module: documentation corrections

Lunik committed:
gitlab_module: python lint

Lunik commented:
@dagwieers Just made the changes (from 0ed1463 to b2f1dd7). What do you this ?

Shaps commented:
@Lunik `module` already has a method called `deprecate` which you can call to raise deprecation warnings ( It basically issues the same warning message, but it lets you say in which version the parameter will be removed ). Have a look here: [ec2_instance](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/ec2_instance.py#L1618)

Lunik committed:
gitlab_module: deprecate options and aliases for ansible 2.10

Lunik commented:
Should we worry about #51851 ?

Shaps commented:
@Lunik well, yeah, I guess whichever PR gets merged after will have to rebase the changes. How far away from merging this PR do you think we are?

Lunik commented:
@Shaps On this PR, we only need review now.

Shaps commented on lib/ansible/modules/source_control/gitlab_deploy_key.py:
```diff
@@ -1,91 +1,95 @@
-    access_token: "{{ access_token }}"
+- name: "Remove the previous deploy key from the project"
+  gitlab_deploy_key:
+    server_url: https://gitlab.example.com/api
```
This needs to be updated with `api_url`

Lunik replied:
Fixed with c5d837e

Shaps commented on lib/ansible/modules/source_control/gitlab_deploy_key.py:
```diff
@@ -94,140 +98,226 @@
+        self.deployKeyObject = deployKey
+        if changed:
+            if self._module.check_mode:
+                self._module.exit_json(changed=True, msg="Deploy_key should have been updated.")
```
I would say `Deploy key has been updated` instead of `should have been updated`

Lunik replied:
Fixed with f261659

Shaps commented on lib/ansible/modules/source_control/gitlab_deploy_key.py:
```diff
@@ -94,140 +98,226 @@
+
+    def deleteDeployKey(self):
+        if self._module.check_mode:
+            self._module.exit_json(changed=True, msg="Deploy_key should have been deleted.")
```
Same as for the updated, I'd rather say `has been deleted`

Lunik replied:
Fixed with f261659

Shaps commented on lib/ansible/modules/source_control/gitlab_deploy_key.py:
```diff
@@ -94,140 +98,226 @@
+        return self.deployKeyObject.delete()
+
+
+def depreaction_warning(module):
```
Typo in `deprecation`

Lunik replied:
Fixed with c271137

Shaps commented on lib/ansible/modules/source_control/gitlab_deploy_key.py:
```diff
@@ -94,140 +98,226 @@
-    access_token = module.params['access_token']
-    private_token = module.params['private_token']
-    project = module.params['project']
+    depreaction_warning(module)
```
Same typo

Lunik replied:
Fixed with c271137

Shaps commented on lib/ansible/modules/source_control/gitlab_deploy_key.py:
```diff
@@ -94,140 +98,226 @@
 
-    if not access_token and not private_token:
-        module.fail_json(msg="need either access_token or private_token")
+    module.exit_json(changed=True, msg="Deploy_key should have been deleted.")
```
Why is this in here?

Lunik replied:
It shouldn't be...

Shaps commented on lib/ansible/modules/source_control/gitlab_deploy_key.py:
```diff
@@ -94,140 +98,226 @@
-        module.fail_json(msg="need either access_token or private_token")
+    module.exit_json(changed=True, msg="Deploy_key should have been deleted.")
+    if not HAS_GITLAB_PACKAGE:
+        module.fail_json(msg="Missing required gitlab module (check docs or install with: pip install python-gitlab")
```
use the `missing_required_lib` function from `ansible.module_utils.basic`

Lunik replied:
Fixed with 8bda23d

Shaps commented on lib/ansible/modules/source_control/gitlab_group.py:
```diff
@@ -15,59 +17,79 @@
 
 EXAMPLES = '''
 - name: "Delete Gitlab Group"
   local_action:
```
Let's try not using `local_action` in the examples, just use `gitlab_group` ( applies to all subsequent examples )

Lunik replied:
Fixed by 7ddb556

Shaps commented on lib/ansible/modules/source_control/gitlab_group.py:
```diff
@@ -76,142 +98,283 @@
         if changed:
             if self._module.check_mode:
-                self._module.exit_json(changed=True, result="Group should have updated.")
+                self._module.exit_json(changed=True, msg="Group should have been updated.")
```
Same as for other modules, use `Group has been updated`

Lunik replied:
Fixed with f261659

Shaps commented on lib/ansible/modules/source_control/gitlab_group.py:
```diff
@@ -76,142 +98,283 @@
+        return False
+
+
+def depreaction_warning(module):
```
typo in `deprecation`

Lunik replied:
Fixed with c271137

Shaps commented on lib/ansible/modules/source_control/gitlab_group.py:
```diff
@@ -76,142 +98,283 @@
 def main():
+    argument_spec = basic_auth_argument_spec()
+    argument_spec.update(dict(
+        server_url=dict(type='str', required=True, removed_in_version=2.10),
```
Shouldn't it be `api_url`?

Lunik replied:
options `api_url` is imported from `basic_auth_argument_spec`. We need to keep option `server_url` for now to keep backward compatibility. Those options will be deprecated in the future

Shaps commented on lib/ansible/modules/source_control/gitlab_hooks.py:
```diff
@@ -127,174 +122,282 @@
+        self.hookObject = hook
+        if changed:
+            if self._module.check_mode:
+                self._module.exit_json(changed=True, msg="Hook should have been updated.")
```
`Hook has been updated`

Lunik replied:
Fixed with f261659

Shaps commented on lib/ansible/modules/source_control/gitlab_hooks.py:
```diff
@@ -127,174 +122,282 @@
-            return False
-    return True
+
+def depreaction_warning(module):
```
Typo in `deprecation`

Lunik replied:
Fixed with c271137

Shaps commented on lib/ansible/modules/source_control/gitlab_hooks.py:
```diff
@@ -127,174 +122,282 @@
+    hook_token = module.params['token']
+
+    if not HAS_GITLAB_PACKAGE:
+        module.fail_json(msg="Missing required gitlab module (check docs or install with: pip install python-gitlab")
```
use `missing_required_lib` from `ansible.module_utils.basic`

Lunik replied:
Fixed with 8bda23d

Shaps commented on lib/ansible/modules/source_control/gitlab_project.py:
```diff
@@ -138,257 +142,274 @@
-            owner = result['namespace']['name']
-            if owner == project_owner:
-                return self._gitlab.deleteproject(result['id'])
+                self._module.exit_json(changed=True, msg="Project should have been updated.")
```
`Project has been updated`

Lunik replied:
Fixed with f261659

Shaps commented on lib/ansible/modules/source_control/gitlab_project.py:
```diff
@@ -138,257 +142,274 @@
+    '''
+    def createProject(self, namespace, arguments):
+        if self._module.check_mode:
+            self._module.exit_json(changed=True, msg="Project should have been created.")
```
`Project has been created`

Lunik replied:
Fixed with f261659

Shaps commented on lib/ansible/modules/source_control/gitlab_project.py:
```diff
@@ -138,257 +142,274 @@
+
+    def deleteProject(self):
+        if self._module.check_mode:
+            self._module.exit_json(changed=True, msg="Project should have been deleted.")
```
`Project has been deleted`

Lunik replied:
Fixed with f261659

Shaps commented on lib/ansible/modules/source_control/gitlab_project.py:
```diff
@@ -138,257 +142,274 @@
 
-    def getUserId(self, user_name):
-        user_data = self._gitlab.getusers(search=user_name)
+def depreaction_warning(module):
```
Typo in `deprecation`

Lunik replied:
Frixed with c271137

Shaps commented on lib/ansible/modules/source_control/gitlab_project.py:
```diff
@@ -138,257 +142,274 @@
-    else:
-        module.fail_json(msg="No login credentials are given. Use login_user with login_password, or login_token")
+    if not HAS_GITLAB_PACKAGE:
+        module.fail_json(msg="Missing required gitlab module (check docs or install with: pip install python-gitlab")
```
Use `missing_required_lib` from `ansible.module_utils.basic`

Lunik replied:
Fixed with 8bda23d

Shaps commented on lib/ansible/modules/source_control/gitlab_runner.py:
```diff
@@ -30,414 +30,357 @@
+        self.runnerObject = runner
+        if changed:
+            if self._module.check_mode:
+                self._module.exit_json(changed=True, msg="Runner should have been updated.")
```
`Runner has been updated`

Lunik replied:
Fixed with f261659

Shaps commented on lib/ansible/modules/source_control/gitlab_runner.py:
```diff
@@ -30,414 +30,357 @@
+    '''
+    def createRunner(self, arguments):
+        if self._module.check_mode:
+            self._module.exit_json(changed=True, msg="Runner should have been created.")
```
`Runner has been created`

Lunik replied:
Fixed with f261659

Shaps commented on lib/ansible/modules/source_control/gitlab_runner.py:
```diff
@@ -30,414 +30,357 @@
+        return runner.delete()
+
+
+def depreaction_warning(module):
```
Typo in `deprecation`

Lunik replied:
Fixed c271137

Shaps commented on lib/ansible/modules/source_control/gitlab_runner.py:
```diff
@@ -30,414 +30,357 @@
+    registration_token = module.params['registration_token']
+
+    if not HAS_GITLAB_PACKAGE:
+        module.fail_json(msg="Missing required gitlab module (check docs or install with: pip install python-gitlab")
```
Use `missing_required_lib` from `ansible.module_utils.basic`

Lunik replied:
Fixed with 8bda23d

Shaps commented on lib/ansible/modules/source_control/gitlab_user.py:
```diff
@@ -116,233 +146,380 @@
-
-    def deleteUser(self, user_username):
-        user_id = self.getUserId(user_username)
+                self._module.exit_json(changed=True, msg="User should have beenupdated.")
```
`User has been updated`

Lunik replied:
Fixed with f261659

Shaps commented on lib/ansible/modules/source_control/gitlab_user.py:
```diff
@@ -116,233 +146,380 @@
+    '''
+    def createUser(self, arguments):
+        if self._module.check_mode:
+            self._module.exit_json(changed=True, msg="User should have been created.")
```
`User has been created`

Lunik replied:
Fixed with f261659

Shaps commented on lib/ansible/modules/source_control/gitlab_user.py:
```diff
@@ -116,233 +146,380 @@
-                return group['id']
+    def deleteUser(self):
+        if self._module.check_mode:
+            self._module.exit_json(changed=True, msg="User should have been deleted.")
```
`User has been deleted`

Lunik replied:
Fixed with f261659

Shaps commented on lib/ansible/modules/source_control/gitlab_user.py:
```diff
@@ -116,233 +146,380 @@
-            self._module.exit_json(changed=True, result="The user %s is updated" % user_username)
-        else:
-            self._module.exit_json(changed=False, result="The user %s is already up2date" % user_username)
+def depreaction_warning(module):
```
Typo in `deprecation`

Lunik replied:
Fixed with c271137

Shaps commented on lib/ansible/modules/source_control/gitlab_user.py:
```diff
@@ -116,233 +146,380 @@
+    user_external = module.params['external']
+
+    if not HAS_GITLAB_PACKAGE:
+        module.fail_json(msg="Missing required gitlab module (check docs or install with: pip install python-gitlab")
```
Use `missing_required_lib` from `ansible.module_utils.basic`

Lunik replied:
Fixed with 8bda23d

Shaps commented:
@Lunik I have added some review comments, I tried highlighting all the lines that require change, although I may very well have missed some duplicate changes, make sure you check for the changes you see more than once in all the modules

dagwieers commented:
Reminder, check_mode should behave identical to a normal run, which means that we don't return a different message because someone ran the module in check_mode. We only do not make the change, but we do everything else exactly as during a normal run.

Lunik committed:
gitlab_group: don't use 'local_action' is documentation example

Lunik committed:
gitlab_module: correct return messages

Lunik committed:
gitlab_module: use module_util 'missing_required_lib' when python library is missing

Lunik committed:
gitlab_module: fix typo in function name.

Lunik committed:
gitlab_modules: unify return msg on check_mode

Lunik committed:
gitlab_modules: don't use deprecated options in examples

Lunik commented:
> Reminder, check_mode should behave identical to a normal run, which means that we don't return a different message because someone ran the module in check_mode. We only do not make the change, but we do everything else exactly as during a normal run.

Should be ok with c5d837e

Lunik commented:
ready_for_review

dagwieers commented:
Great work, lots of things fixed !
